### PR TITLE
React upload wizard required fields & validation (SCP-3740)

### PR DESCRIPTION
**Pull request by devonbush**
This adds the validations detailed in the [google doc](https://docs.google.com/document/d/1DEWrROU-zgovNaK9lR_SnshCf7tmCdBoz3TihRsgWCM/edit#heading=h.29woqp4kz24g).  

Name uniqueness and file extension errors are shown directly under the file chooser
image
image
(note the latter screenshot is actually difficult to get, since for the most part you are prevented from even  selecting files without the approved extensions.  But for some weird reason the OS X file chooser lets you pick '.js' files no matter what types you pass in? 

Required fields are indicated with an asterisk next to the field label, and the save button displays a tooltip on hover/focus indicating any missing required fields (as well as repeating any file name / extension errors)
image

This also includes the start to what I hope will become good automated test coverage of the wizard frontend.  I've created more jira tickets for large areas needing additional automation coverage.

TO TEST:
  1. Create a new study (with react upload wizard feature flag on)
  2. attempt to save the empty raw counts file that appears
  3. confirm the validation errors appear
  3. fill out the raw counts form, selecting a valid file
  4. confirm you can save the raw counts file
  5. move to processed matrix step.
  6. Attempt to select the same file you uploaded for raw counts
  7. confirm a 'that file already exists in your study' message appears
  8. Move through the other steps, confirming that file validation and required field validation is present

### CODE REVIEW

File: app/javascript/components/upload/CoordinateLabelStep.js [hunk 1]
```diff
@@ -70,23 +69,26 @@ function CoordinateLabelForm({
...
             ...
           </pre>
-        <p>
-          <strong>These are not cluster files</strong> - they are annotations to overlay on top of a cluster.<br/>
-          The file must be a plain text (.txt) file with at least 3 columns and a header row containing the values <strong>X</strong>, <strong>Y</strong>, and <strong>LABELS</strong>.  The file may have an optional column of <strong>Z</strong> (for 3d clusters).
-          The last column must contain text labels to display at the specified coordinates.
-        </p>
-        <p><i className="fas fa-fw fa-exclamation-triangle text-warning"></i> The coordinates of the labels must fall inside the ranges of the cluster they are associated with for them to render.</p>
+          <p>
+            <strong>These are not cluster files</strong> - they are annotations to overlay on top of a cluster.<br/>
```
Comment (bistline) [inline app/javascript/components/upload/CoordinateLabelStep.js:73]: Realizing this language is a bit ambiguous.
```suggestion
            <strong>These are not cluster files</strong> - they are annotations to overlay on top of a scatter plot.<br/>
```

File: app/javascript/components/upload/ExpressionFileForm.js [hunk 2]
```diff
@@ -5,54 +5,67 @@ import Select from 'lib/InstrumentedSelect'
 import MTXBundledFilesForm from './MTXBundledFilesForm'
 import FileUploadControl, { FileTypeExtensions } from './FileUploadControl'
 import { TextFormField, SavingOverlay, SaveDeleteButtons } from './form-components'
+import { findBundleChildren, validateFile } from './upload-utils'
+
+const REQUIRED_FIELDS = [{ label: 'species', propertyName: 'taxon_id' },
```
Comment (bistline) [inline app/javascript/components/upload/ExpressionFileForm.js:10]: For now I understand that these are needed to control form logic for each individual form, but this just seems like an easy place for things to get out of sync with the Ruby models when we make changes.  A future enhancement could be an API endpoint that would serve this info.  We do have an existing one that might be good candidate/starting point (at `api/v1/schemas/study_files`) - we could pass in the `file_type` and get back the required fields.  It currently doesn't have any info right now about `expression_file_info`, so we'd need to add in that logic as well.

File: test/js/upload-wizard/file-info-responses.js [hunk 3]
```diff
@@ -0,0 +1,293 @@
...
+    'x_axis_max': null,
+    'x_axis_min': null,
+    'y_axis_label': '',
+    'y_axis_max': null,
+    'y_axis_min': null,
+    'z_axis_label': '',
+    'z_axis_max': null,
+    'z_axis_min': null,
+    'upload': {
+      'url': '/Users/dbush/code/scp/single_cell_portal_core/data/51e9c0d33e9b698b118bbb884fab092a1b2d22a9f4133966010560c60835361a/60a2b9fbcc7ba082358b544a/original/metadata.txt'
```
Comment (bistline) [inline test/js/upload-wizard/file-info-responses.js:53]: Nit - It doesn't seem to impact the tests, so can we change these urls?  Seeing a local path from your machine makes my code smell radar go off.

File: test/js/upload-wizard/validations.test.js [hunk 4]
```diff
@@ -0,0 +1,212 @@
...
+      status: 'new',
+      _id: '10'
+    }
+    const allFiles = [{ file }]
+    const msgs = validateFile({
+      file, allFiles, allowedFileTypes: ['.txt', '.txt.gz'],
+      requiredFields: [{ label: 'species', propertyName: 'taxon_id' }]
+    })
+    expect(msgs.taxon_id).toEqual('You must specify species')
+    expect(msgs.name).toEqual('You must specify name')
```
Comment (bistline) [inline test/js/upload-wizard/validations.test.js:174]: I'm getting a JS test failure at this point, but for some reason builds aren't running in Travis for this PR.  Not sure what the issue is.
```
Summary of all failing tests
 FAIL  test/js/upload-wizard/validations.test.js (12.178 s)
  ● it checks presence of required fields › fails if required fields are not present

    expect(received).toEqual(expected) // deep equality

    Expected: "You must specify name"
    Received: undefined

      172 |     })
      173 |     expect(msgs.taxon_id).toEqual('You must specify species')
    > 174 |     expect(msgs.name).toEqual('You must specify name')
          |                       ^
      175 |   })
      176 |
      177 |   it('messages both if multiple required fields are not present', async () => {

      at Object.<anonymous> (test/js/upload-wizard/validations.test.js:174:23)
```

**Review by bistline — APPROVED** [review]
On the whole, this looks good.  I noted a place for a future enhancement re: required fields for study files.  Also, I noticed two minor UX papercuts:

After creating a new study, for some reason the modal isn't html encoding the alert:
Screen Shot 2021-10-05 at 10 09 16 AM



Also, after I upload a file, and then try to upload the same file, _all_ of the forms now show error messages for the file, including forms that don't have any files yet.  If possible, it would be great to scope the error validation to only the form that is being used:
Screen Shot 2021-10-05 at 10 11 30 AM

File: app/javascript/components/upload/ExpressionFileForm.js [hunk 5]
```diff
@@ -5,54 +5,67 @@ import Select from 'lib/InstrumentedSelect'
...
+
   const associatedRawCounts = file.expression_file_info.raw_counts_associations.map(id => ({
     label: rawCountsOptions.find(rf => rf.value == id)?.label,
     value: id
   }))
 
   return <div className="row top-margin" key={file._id}>
     <div className="col-md-12">
-      <form id={`clusterForm-${file._id}`}
+      <form id={`expForm-${file._id}`}
```
Comment (eweitz) [inline app/javascript/components/upload/ExpressionFileForm.js:46]: Let's use hyphen-case for attribute values, per convention:

```suggestion
      <form id={`exp-form-${file._id}`}
```

File: app/javascript/components/upload/ExpressionFileForm.js [hunk 6]
```diff
@@ -5,54 +5,67 @@ import Select from 'lib/InstrumentedSelect'
...
           updateFile={updateFile}
-          allowedFileTypes={isMtxFile ? FileTypeExtensions.mtx : FileTypeExtensions.plainText}/>
+          validationMessages={validationMessages}
+          allowedFileTypes={allowedFileTypes}/>
         <div className="form-group">
           <label>Matrix file type:</label><br/>
           <label className="sublabel">
             <input type="radio"
-              name={`rawCountsType-${file._id}`}
+              name={`expMatrixType-${file._id}`}
```
Comment (eweitz) [inline app/javascript/components/upload/ExpressionFileForm.js:60]:
```suggestion
              name={`exp-matrix-type-${file._id}`}
```

File: app/javascript/components/upload/ExpressionFileForm.js [hunk 7]
```diff
@@ -5,54 +5,67 @@ import Select from 'lib/InstrumentedSelect'
...
               value="false"
               checked={!isMtxFile}
               onChange={e => updateFile(file._id, { file_type: 'Expression Matrix' })} />
               &nbsp;Expression Matrix
           </label>
           <label className="sublabel">
             <input type="radio"
-              name={`rawCountsType-${file._id}`}
+              name={`expMatrixType-${file._id}`}
```
Comment (eweitz) [inline app/javascript/components/upload/ExpressionFileForm.js:68]:
```suggestion
              name={`exp-matrix-type-${file._id}`}
```

**Review by eweitz — APPROVED** [review]

File: app/javascript/components/upload/ExpressionFileForm.js (hunk 2 above)
Comment (devonbush) [reply app/javascript/components/upload/ExpressionFileForm.js:10]: yup

**Pull request merged**